main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -2548,6 +2548,23 @@ object Typer {
           elementTypes <- traverseM(elms)(visit)
         } yield Type.mkTuple(elementTypes, loc)
 
+
+      case KindedAst.Pattern.Record(pats, pat, tvar, loc) =>
+        val freshRowVar = Type.freshVar(Kind.RecordRow, loc.asSynthetic)
+        val freshRecord = Type.mkRecord(freshRowVar, loc.asSynthetic)
+        for {
+          optRecordTail <- traverseOptM(pat)(visit)
+          emptyRecord = Type.mkRecord(Type.mkRecordRowEmpty(loc.asSynthetic), loc.asSynthetic)
+          recordTail = optRecordTail.getOrElse(emptyRecord)
+          _recordExtension <- unifyTypeM(freshRecord, recordTail, loc.asSynthetic)
+          ps <- traverseM(pats)(visitRecordFieldPattern(_, root))
+          patTypes = ps.foldRight(freshRowVar: Type) {
+            case ((f, t, l), acc) => Type.mkRecordRowExtend(f, t, acc, l)
+          }
+          resultType = Type.mkRecord(patTypes, loc)
+          _ <- unifyTypeM(resultType, tvar, loc)
+        } yield resultType
+
```

Comment:
I think the above can be improved a bit in terms of readability:

- We can introduce several of the components with `val` bindings before entering the monadic for (i.e. avoid `=` in the for).
- Instead of a fold, a small helper function introduced locally would seem more appropriate.
- `pat` is treated as an option, but can we simply not determine what it is earlier? E.g. either its the empty record or its a variable, but both are types, right?